data/metric/metrics.go

```diff
@@ -82,44 +116,98 @@ func (ms *Set) SetMetric(name string, value interface{}, sourceType SourceType)
 	return strconv.ParseFloat(fmt.Sprintf("%v", value), 64)
 }
 
-func (ms *Set) sample(name string, floatValue float64, sourceType SourceType) (float64, error) {
-	sampledValue := 0.0
+func (ms *Set) elapsedDifference(name string, absolute interface{}, sourceType SourceType) (elapsed float64, err error) {
+	if ms.storer == nil {
+		err = ErrNoStoreToCalcDiff
+		return
+	}
+
+	newValue, err := castToFloat(absolute)
+	if err != nil {
+		err = ErrNonNumeric
+		return
+	}
+
+	// Fetch last value & time
+	var oldValue float64
+	oldTime, err := ms.storer.Get(ms.namespace(name), &oldValue)
+	if err != nil && err != persist.ErrNotFound {
+		return
+	}
 
-	// Retrieve the last value and timestamp from Storer
-	var oldval float64
-	oldTime, err := ms.storer.Get(name, &oldval)
+	// Store new value & time (no IO flush until Save)
+	newTime := ms.storer.Set(ms.namespace(name), newValue)
+
+	// First value
 	if err == persist.ErrNotFound {
-		oldval = 0
-	} else if err != nil {
-		return sampledValue, errors.Wrapf(err, "sample-key: %s", name)
+		return 0, nil
 	}
-	// And replace it with the new value which we want to keep
-	newTime := ms.storer.Set(name, floatValue)
 
-	if err == nil {
-		duration := newTime - oldTime
-		if duration == 0 {
-			return sampledValue, fmt.Errorf("samples for %s are too close in time, skipping sampling", name)
-		}
+	// Time constraints
+	duration := newTime - oldTime
+	if duration == 0 {
+		err = ErrTooCloseSamples
+		return
+	}
 
-		if floatValue-oldval < 0 {
-			return sampledValue, fmt.Errorf("source for %s was reseted, skipping sampling", name)
-		}
-		if sourceType == DELTA {
-			sampledValue = floatValue - oldval
-		} else {
-			sampledValue = (floatValue - oldval) / float64(duration)
-		}
+	elapsed = newValue - oldValue
+	if elapsed < 0 {
+		err = ErrNegativeDiff
+		return
+	}
+
+	if sourceType == RATE {
+		elapsed = elapsed / float64(duration)
+	}
+
+	return
+}
+
+// prefix a metric name with a namespace based on the alphabetical order of the set related attributes.
+func (ms *Set) namespace(metricName string) string {
+	ns := ""
+	separator := ""
+
+	attrs := ms.nsAttributes
+	sort.Sort(Attributes(attrs))
+
+	for _, attr := range attrs {
+		ns = fmt.Sprintf("%s%s%s", ns, separator, attr.Namespace())
+		separator = NSSeparator
 	}
 
-	return sampledValue, nil
+	return fmt.Sprintf("%s%s%s", ns, separator, metricName)
+}
+
+// Namespace generates the string value of an attribute used to namespace a metric.
+func (a *Attribute) Namespace() string {
+	return fmt.Sprintf("%s%s%s", a.Key, NSAttributeSeparator, a.Value)
 }
 
 // MarshalJSON adapts the internal structure of the metrics Set to the payload that is compliant with the protocol
 func (ms Set) MarshalJSON() ([]byte, error) {
 	return json.Marshal(ms.Metrics)
 }
+
+// Required for Go < v.18, as these do not include sort.Slice
+
+// Attributes list of attributes
+type Attributes []Attribute
+
+// Len ...
+func (a Attributes) Len() int { return len(a) }
```

Comment:
What are `Len`, `Swap` and `Less` used for?

Reply:
it's on the comment above: https://github.com/newrelic/infra-integrations-sdk/pull/106/files#diff-83416e4d39b0688a4bec5f4a2c076486R196

> Required for Go < v.18, as these do not include sort.Slice

you can see that change isolated in the commit https://github.com/newrelic/infra-integrations-sdk/pull/106/commits/7da060855a4e41e7ee36a87f451c25ee18e59229

Reply:
Hmm ok. I forgot the implicit implementation of interfaces in Go.